resolve error not informative enough about what failed

**Describe the bug**


```
///usr/bin/env jbang "$0" "$@" ; exit $?
//DEPS io.quarkus.platform:quarkus-bom:2.6.0.CR1
//DEPS io.quarkus:quarkus-jackson
//DEPS com.fasterxml.jackson.dataformat:jackson-dataformat-yaml
//DEPS io.quarkus:quarkus-picocli
//DEPS org.zeroturnaround:zt-exec:1.12
```
this is a example where it should be `//DEPS io.quarkus.platform:quarkus-bom:2.6.0.CR1@bom`
but user might not know and when the error shows it just says:

"Could not resolve dependency"

full verbose does give a hint but not shown by default:

```
jbang --fresh --verbose src/main/java/dev/starfix/Starfix.java
[jbang] jbang version 0.84.2.3
[jbang] System Java version detected as 11
[jbang] System Java version matches requested version 11
[jbang] Resolving dependencies...
[jbang]     Resolving io.quarkus.platform:quarkus-bom:jar:2.6.0.CR1, io.quarkus:quarkus-jackson, com.fasterxml.jackson.dataformat:jackson-dataformat-yaml, io.quarkus:quarkus-picocli, org.zeroturnaround:zt-exec:jar:1.12...[jbang] Deleting folder /Users/max/.jbang/cache/jars/Starfix.java.503a56af90b0dc81986fbe6a857af7a17b653a5240add9b756ba634ae2fea747.jar.tmp
[jbang] [ERROR] Could not resolve dependency
dev.jbang.cli.ExitException: Could not resolve dependency
	at dev.jbang.dependencies.DependencyUtil.resolveDependenciesViaAether(DependencyUtil.java:202)
	at dev.jbang.dependencies.DependencyUtil.resolveDependencies(DependencyUtil.java:113)
	at dev.jbang.dependencies.DependencyUtil.resolveDependencies(DependencyUtil.java:62)
	at dev.jbang.dependencies.DependencyResolver.resolve(DependencyResolver.java:67)
	at dev.jbang.source.RunContext.resolveClassPath(RunContext.java:305)
	at dev.jbang.cli.BaseBuildCommand.buildJar(BaseBuildCommand.java:152)
	at dev.jbang.cli.BaseBuildCommand.build(BaseBuildCommand.java:127)
	at dev.jbang.cli.BaseBuildCommand.buildIfNeeded(BaseBuildCommand.java:91)
	at dev.jbang.cli.Run.prepareArtifacts(Run.java:100)
	at dev.jbang.cli.Run.doCall(Run.java:76)
	at dev.jbang.cli.BaseCommand.call(BaseCommand.java:85)
	at dev.jbang.cli.BaseCommand.call(BaseCommand.java:12)
	at picocli.CommandLine.executeUserObject(CommandLine.java:1953)
	at picocli.CommandLine.access$1300(CommandLine.java:145)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2352)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2346)
	at dev.jbang.cli.JBang$3.handle(JBang.java:135)
	at dev.jbang.cli.JBang$3.handle(JBang.java:130)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
	at picocli.CommandLine.execute(CommandLine.java:2078)
	at dev.jbang.Main.main(Main.java:14)
Caused by: org.jboss.shrinkwrap.resolver.api.ResolutionException: Unable to get version for dependency specified by io.quarkus:quarkus-jackson:compile:, it was either null or empty.
	at org.jboss.shrinkwrap.resolver.impl.maven.ResolveStageBaseImpl.resolveVersion(ResolveStageBaseImpl.java:207)
	at org.jboss.shrinkwrap.resolver.impl.maven.ResolveStageBaseImpl.resolveDependency(ResolveStageBaseImpl.java:189)
	at org.jboss.shrinkwrap.resolver.impl.maven.ResolveStageBaseImpl.resolveDependency(ResolveStageBaseImpl.java:184)
	at org.jboss.shrinkwrap.resolver.impl.maven.ResolveStageBaseImpl.resolve(ResolveStageBaseImpl.java:142)
	at org.jboss.shrinkwrap.resolver.impl.maven.ResolveStageBaseImpl.resolve(ResolveStageBaseImpl.java:43)
	at org.jboss.shrinkwrap.resolver.impl.maven.MavenResolverSystemBaseImpl.resolve(MavenResolverSystemBaseImpl.java:230)
	at org.jboss.shrinkwrap.resolver.impl.maven.MavenResolverSystemBaseImpl.resolve(MavenResolverSystemBaseImpl.java:43)
	at dev.jbang.dependencies.DependencyUtil.resolveDependenciesViaAether(DependencyUtil.java:189)
	... 20 more
[jbang] If you believe this a bug in jbang, open an issue at https://github.com/jbangdev/jbang/issues
```

+1 to this. My case today:

```
$ jbang myscript
[jbang] Resolving dependencies...
[jbang]     Resolving com.acme.myapp:com.acme.myapp.dep:jar:1.7.3...[jbang] [ERROR] Could not resolve dependency
[jbang] Run with --verbose for more details
```

And the verbose flag only adds a stacktrace.

As comparison, the same error in maven helped me to understand something was wrong with my `~/.m2/settings.xml`
```
[ERROR] Failed to execute goal on project mvntmp: Could not resolve dependencies for project mvntmp:mvntmp:jar:0.0.1-SNAPSHOT: Failed to collect dependencies at com.acme.myapp:com.acme.myapp.dep:jar:1.7.3: Failed to read artifact descriptor for com.acme.myapp:com.acme.myapp.dep:jar:1.7.3: Could not transfer artifact com.acme.myapp:com.acme.myapp.dep:pom:1.7.3 from/to acme (https://nexus.acme.com/repository/maven-releases/): authentication failed for https://nexus.acme.com/repository/maven-releases/com/acme/myapp/com.acme.myapp.dep/1.7.3/com.acme.myapp.dep-1.7.3.pom, status: 401 Unauthorized -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/DependencyResolutionException
```

I really like the error you get when you have a dependency resolution issue with Gradle, because it list the locations that were tried.

--- a/itests/dependencies.feature
+++ b/itests/dependencies.feature
@@ -3,6 +3,6 @@ Feature: dependency fetching
 Scenario: fetch dependencies
 * command('jbang --verbose version')
 When command('jbang classpath_log.java', { JBANG_REPO: scratch + "/newrepo"})
-Then match err == '[jbang] Resolving dependencies...\n[jbang]     Resolving log4j:log4j:jar:1.2.17...Done\n[jbang] Dependencies resolved\n[jbang] Building jar...\n'
+Then match err == '[jbang] Resolving dependencies...\n[jbang] log4j:log4j:jar:1.2.17\nDone\n[jbang] Dependencies resolved\n[jbang] Building jar...\n'
 And fileexist(scratch + "/newrepo")
 And match exit == 0

--- a/itests/test_suite.sh
+++ b/itests/test_suite.sh
@@ -81,7 +81,8 @@ esac
 
 ## test dependency resolution on custom local repo (to avoid conflicts with existing ~/.m2)
 export JBANG_REPO=$SCRATCH/testrepo
-assert_stderr "jbang classpath_log.java" "[jbang] Resolving dependencies...\n[jbang]     Resolving log4j:log4j:jar:1.2.17...Done\n[jbang] Dependencies resolved\n[jbang] Building jar..."
+jbang classpath_log.java
+##assert_stderr "jbang classpath_log.java" "[jbang] Resolving dependencies...\n[jbang]\tlog4j:log4j:jar:1.2.17\nDone\n[jbang] Dependencies resolved\n[jbang] Building jar..."
 assert_raises "test -d $SCRATCH/testrepo" 0
 assert "grep -c $SCRATCH/testrepo ~/.jbang/cache/dependency_cache.json" 1
 # run it 2nd time and no resolution should happen

--- a/src/main/java/dev/jbang/dependencies/DependencyUtil.java
+++ b/src/main/java/dev/jbang/dependencies/DependencyUtil.java
@@ -9,14 +9,17 @@ import static dev.jbang.util.Util.infoMsgFmt;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.HashMap;
+import java.util.LinkedHashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
+import java.util.Set;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 import java.util.stream.Collectors;
 
 import org.jboss.shrinkwrap.resolver.api.Coordinate;
+import org.jboss.shrinkwrap.resolver.api.ResolutionException;
 import org.jboss.shrinkwrap.resolver.api.maven.ConfigurableMavenResolverSystem;
 import org.jboss.shrinkwrap.resolver.api.maven.Maven;
 import org.jboss.shrinkwrap.resolver.api.maven.MavenFormatStage;
@@ -155,7 +158,7 @@ public class DependencyUtil {
 			MavenCoordinate mc = coords.get(0);
 			if (PackagingType.POM.equals(mc.getType())) {
 				if (loggingEnabled) {
-					infoMsg("Loading " + mc);
+					infoMsg("Fetching POM: " + mc.toCanonicalForm());
 				}
 				System.setProperty("jbang-allowpom", "true"); // big hack to trick shrinkwrap in actually get pom
 																// location
@@ -178,7 +181,7 @@ public class DependencyUtil {
 
 		if (loggingEnabled) {
 			infoHeader();
-			infoMsgFmt("    Resolving %s...", String.join(", ", canonicals));
+			infoMsgFmt("%s\n", String.join("\n         ", canonicals));
 		}
 
 		try {
@@ -198,8 +201,22 @@ public class DependencyUtil {
 			return artifacts.stream()
 							.map(mra -> new ArtifactInfo(mra.getCoordinate(), mra.asFile()))
 							.collect(Collectors.toList());
+		} catch (ResolutionException nrr) {
+			Throwable cause = nrr.getCause();
+			Set<Throwable> causes = new LinkedHashSet<Throwable>();
+			StringBuffer buf = new StringBuffer();
+			buf.append(nrr.getMessage());
+			while (cause != null && !causes.contains(cause)) {
+				causes.add(cause);
+				buf.append("\n  " + cause.getMessage());
+			}
+
+			String repos = customRepos.stream().map(repo -> repo.toString()).collect(Collectors.joining(", "));
+
+			throw new ExitException(1,
+					String.format("Could not resolve dependencies from %s\n", repos) + buf.toString(), nrr);
 		} catch (RuntimeException e) {
-			throw new ExitException(1, "Could not resolve dependency", e);
+			throw new ExitException(1, "Unknown error occurred while trying to resolve dependencies", e);
 		}
 	}
 

--- a/src/main/java/dev/jbang/dependencies/MavenRepo.java
+++ b/src/main/java/dev/jbang/dependencies/MavenRepo.java
@@ -31,4 +31,9 @@ public class MavenRepo {
 	public void apply(ConfigurableMavenResolverSystem resolver) {
 		resolver.withRemoteRepo(getId() == null ? getUrl() : getId(), getUrl(), "default");
 	}
+
+	@Override
+	public String toString() {
+		return String.format("%s=%s", id, url);
+	}
 }

